rfcs/12-primitive-shapes.md

```diff
@@ -0,0 +1,545 @@
+
+Meshing could be naturally extended with other libraries or parameters. For example, a sphere by default might use `Icosphere`, but could be extended to generate UV spheres or quad spheres. For 2d types, we could use a crate such as `lyon` to generate 2d meshes from parameterized primitive shapes within the `Meshable` interface.
+
+### Bounding
+
+For use in spatial queries, broad phase collision detection, and raycasting, hypothetical `Bounding` and `Bounding2d` traits are provided. These traits are implemented for types that define position in addition to the shape and size defined in the base primitive. The basic functionality of this trait is to check whether one bounding shape is contained within another bounding shape.
+
+### Colliders
+
+Colliders can provide intersection checks against other colliders, as well as check if they are within a bounding volume.
+
+### 3D and 2D
+
+This RFC provides independent 2d and 3d primitives. Recall that the purpose of this is to provide lightweight types, so there are what appear to be duplicates in 2d and 3d, such as `Line` and `Line2d`. Note that the 2d version of a line is smaller than its 3d counterpart because it is only defined in 2d. 3d geometry (or 2d with depth) is assumed to be the default for most cases. The names of the types were chosen with this in mind.
+
+## Implementation strategy
+
+### Helper Types
+
+```rust
+/// Stores an angle in radians, and supplies builder functions to prevent errors (from_radians, from_degrees)
+struct Angle(f32);
+```
+### Traits (REFERENCE ONLY!)
+
+**These traits are provided as reference to illustrate how these primitive shape types might be used. The details of implementation and interface should be determined in a separate RFC, PR, or independent prototypes.**
+
+```rust
+trait Meshable{
+  fn mesh(&self) -> Mesh;
+};
+
+trait Bounding {
+  fn within(&self, other: &impl Bounding) -> bool;
+  fn contains(&self, collider: &impl Collider) -> bool;
+}
+
+trait Bounding2d {
+  fn within(&self, other: &impl Bounding2d) -> bool;
+  fn contains(&self, collider: &impl Collider2d) -> bool;
+}
+
+trait Collider {
+  fn collide(&self, other: &impl Collider) -> Option(Collision);
+  fn within(&self, bounds: &impl Bounding) -> bool;
+}
+
+trait Collider2d {
+  fn collide(&self, other: &impl Collider2d) -> Option(Collision2d);
+  fn within(&self, bounds: &impl Bounding2d) -> bool;
+}
+```
+
+### 3D Geometry Types
+
+```rust
+struct Point(Vec3)
+
+/// Vector direction in 3D space that is guaranteed to be normalized through its getter/setter.
+struct Direction(Vec3)
+impl Meshable for Direction {}
+
+struct Plane {
```

Comment:
I use just 4 `f32`s to define a plane - an `n: Vec3` for direction (which doesn't need to be normalized), and a `d: f32` for the dot product, so that the plane is defined as all points `x: Vec3` such that `n.dot(x) == d`.

If you have a defined center, you can also normalize to `d = 1`, which can be used to define polyhedra using half planes.